### PR TITLE
embeddings from scratch training

### PR DESCRIPTION
**Pull request by anhad13**

### CODE REVIEW

File: src/models.py [hunk 1]
```diff
@@ -237,13 +237,17 @@ def build_embeddings(args, vocab, tasks, pretrained_embs=None):
             d_word = word_embs.size()[1]
             log.info("\tUsing pre-trained word embeddings: %s",
                      str(word_embs.size()))
+        elif args.word_embs=="scratch":
+            log.info("\tScratch Word embs!")
+            d_word=args.d_word
+            word_embs= nn.Embedding(n_token_vocab, d_word).weight
```
Comment (sleepinyourhat) [inline src/models.py]: Are you sure this learns? I don't think I've seen .weight used this way...

File: src/models.py [hunk 2]
```diff
@@ -237,13 +237,17 @@ def build_embeddings(args, vocab, tasks, pretrained_embs=None):
...
         else:
```
Comment (sleepinyourhat) [inline src/models.py]: What does this clause do? It's a serious problem to have two different options that both claim to be learning embeddings from scratch.

Comment (anhad13) [reply src/models.py]: Fixed

File: src/models.py (hunk 1 above)
Comment (anhad13) [reply src/models.py]: Yep, check that it learns and grad is present

**Comment by sleepinyourhat** [comment]
Ready for review?

**Comment by anhad13** [comment]
Yes, this is ready for review

File: src/models.py [hunk 3]
```diff
@@ -237,13 +237,15 @@ def build_embeddings(args, vocab, tasks, pretrained_embs=None):
             d_word = word_embs.size()[1]
             log.info("\tUsing pre-trained word embeddings: %s",
                      str(word_embs.size()))
+        elif args.word_embs=="scratch":
+            log.info("\tScratch Word embs!")
```
Comment (sleepinyourhat) [inline src/models.py]: Mind clarifying this (i.e., using a full sentence)?

File: src/preprocess.py [hunk 4]
```diff
@@ -287,7 +287,7 @@ def build_tasks(args):
 
     # 3) build / load word vectors
     word_embs = None
-    if args.word_embs != 'none':
+    if args.word_embs != 'none' and args.word_embs!='scratch':
```
Comment (sleepinyourhat) [inline src/preprocess.py]: Fix style.

File: src/models.py [hunk 5]
```diff
@@ -237,13 +237,15 @@ def build_embeddings(args, vocab, tasks, pretrained_embs=None):
...
+            word_embs= nn.Embedding(n_token_vocab, d_word).weight
         else:
-            log.info("\tLearning word embeddings from scratch!")
-            word_embs = None
-            d_word = args.d_word
+            raise Exception('Not a valid type of word emb. Set to none for elmo.')
 
         embeddings = Embedding(num_embeddings=n_token_vocab, embedding_dim=d_word,
-                               weight=word_embs, trainable=False,
+                               weight=word_embs, trainable=True if args.word_embs=="scratch" else False,
```
Comment (iftenney) [inline src/models.py]: nit: maybe just have a separate argument `train_embeddings` for consistency with other modules (like OpenAI transformer and BERT)?

**Comment by sleepinyourhat** [comment]
Any progress here?

File: src/models.py (hunk 5 above)
Comment (anhad13) [reply src/models.py]: Fixed.

File: src/preprocess.py (hunk 4 above)
Comment (anhad13) [reply src/preprocess.py]: Fixed.

File: src/models.py (hunk 3 above)
Comment (anhad13) [reply src/models.py]: Fixed.

**Comment by anhad13** [comment]
Fixed changes. Ready to merge?

File: config/cola.conf [hunk 6]
```diff
@@ -0,0 +1,23 @@
+include "defaults.conf"
```
Comment (sleepinyourhat) [inline config/cola.conf]: This doesn't look like it should be part of this commit. Is it here intentionally?

File: config/cola.conf [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+JIANT_DATA_DIR="/Users/anhadmohananey/Desktop/NLPResearch/jiant/jiant/data"
```
Comment (sleepinyourhat) [inline config/cola.conf]: This shouldn't live in a conf file, even on your local machine. Use path_config.sh to set these.

File: config/cola.conf [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+// Three main stages of operation
+do_pretrain = 1 // Train the shared sentence encoder model (and task-specific model parameters) on the pretraining tasks in
+              // pretrain_tasks.
+do_target_task_training = 1  // After do_pretrain, train the task-specific model parameters on the target tasks in target_tasks.
+do_full_eval = 1     // Evaluate the model on the tasks on target_tasks.
+
+patience=5
+lr_patience=5
+allow_reuse_of_pretraining_parameters = 1
+exp_name="blah"
```
Comment (sleepinyourhat) [inline config/cola.conf]: Be consistent in style (spacing), and use descriptive names.

File: config/defaults.conf [hunk 9]
```diff
@@ -398,6 +398,8 @@ qqp-alt_pair_attn = 1
 qqp-alt_val_interval = 1000
 qqp-alt_lr = 0.0003
 
+embeddings_train  = 0
```
Comment (sleepinyourhat) [inline config/defaults.conf]: Add a comment explaining this, and move it up next to the other options that involve embeddings.

File: src/models.py (hunk 1 above)
Comment (sleepinyourhat) [reply src/models.py]: Got it. Fix style, though!

File: config/cola.conf (hunk 6 above)
Comment (anhad13) [reply config/cola.conf]: Oh I actually I added the cola.conf file by mistake. I'll remove it from the commit.

File: config/defaults.conf (hunk 9 above)
Comment (anhad13) [reply config/defaults.conf]: Done!

**Comment by anhad13** [comment]
Fixed issues. Removed cola.conf.

File: src/models.py [hunk 10]
```diff
@@ -237,13 +237,15 @@ def build_embeddings(args, vocab, tasks, pretrained_embs=None):
...
+        elif args.word_embs=="scratch":
+            log.info("\tTraining word embeddings from scratch.")
             d_word = args.d_word
+            word_embs = nn.Embedding(n_token_vocab, d_word).weight
+        else:
+            raise Exception('Not a valid type of word emb. Set to none for elmo.')
 
         embeddings = Embedding(num_embeddings=n_token_vocab, embedding_dim=d_word,
-                               weight=word_embs, trainable=False,
+                               weight=word_embs, trainable=(args.embeddings_train==1),
```
Comment (sleepinyourhat) [inline src/models.py]: Do fix the other style issues—you really need to use autopep8 or something similar.

Comment (anhad13) [reply src/models.py]: Thanks, point noted!

**Comment by anhad13** [comment]
PEP8 compliance fixes on src/models.py

**Pull request merged**